PublicMapping/districtbuilder
# Utilize all available memory on a container instance

## Overview

Even with 64 GB of memory on the `r5.2xlarge`, we can't load the new regions configured for production. This pull request changes the default `container_instance_type` to an `r5.4xlarge`, removes the hard CPU limit, and expands the hard memory limit to utilize the EC2 instance as possible. 

See: https://azavea.slack.com/archives/C6UD5FG3T/p1628872715113500

**Update:**

A lot has transpired since I opened this pull request, but I think its purpose is still genuine: to remove the limitations that prevented us from scaling. We determined a few, interesting things:

1) Despite removing CPU and memory limitations, and placing the task on an instance with 128 GB of memory, the task was dying once it hit ~17 GB of memory.
2) I found a few posts on the internet, [like this Stack Overflow question](https://stackoverflow.com/questions/67280057/fatal-error-scavenger-semi-space-copy-allocation-failed), where other folks were running into garbage collection errors at the same memory usage as us (~17 GB).
3) I had a feeling that this was a kernel limitation vs. anything on the application side for a few reasons:
   1) We're storing a **very large** data set in memory.
   2) Cole shared a [blog post](https://deepu.tech/memory-management-in-v8/) on V8 memory management that said:
      > **Large object space:** This is where objects which are larger than the size limits of other spaces live. Each object gets its own [`mmap`](https://en.wikipedia.org/wiki/Mmap)'d region of memory. Large objects are never moved by the garbage collector.
   3) 17 GB is an odd value. It's not a limitation for anything relating to V8 itself. And it didn't feel right in the context of things like map key limits, etc.
4) After tuning `vm.max_map_count` to an arbitrarily high value, the application was able to load all the new census data, consuming a maximum of around 27 GB of memory.

After all this, I persisted `vm.max_map_count` as a less-arbitrary value and scaled us back down to a 64 GB instance type. We're still not in the green, as we've been running into new issues with our database driver (e.g., #950), but this pull request puts us in a much better state.

### Checklist


## Testing Instructions

This branch is what @maurizi and I used to deploy the new census data, so we're testing the Terraform changes in production as we speak.

Using the last SHA we deployed to staging, make sure you can apply the Terraform plan:

<details>

```console
docker-compose -f docker-compose.ci.yml run --rm terraform
export GIT_COMMIT=67d842f
./scripts/infra plan
./scripts/infra apply
```
</details>

Open a shell on one of the container instances and see that `vm.max_map_count` is much higher than the default of `65536`:

<details>

```console
ssh districtbuilder-stg
[ec2-user@ip-10-0-0-38 ~]$ ssh 10.0.1.96
[ec2-user@ip-10-0-1-96 ~]$ cat /proc/sys/vm/max_map_count
192512
```

</details>

See that application health checks are passing:

<details>

```console
http https://app.staging.districtbuilder.org/healthcheck/
```

TBD.

</details>

See that the ECS service is stable:

<details>

TBD.

</details>
